### DISCUSSION TITLE
What is the motivation to change from netstandard2.0 to netstandard2.1?

### DISCUSSION THREAD

**Discussion by Karql** [LuckyPennySoftware/AutoMapper#3873] [NA]
@lbargaoanu What is the motivation to change from netstandard2.0 to netstandard2.1?
In this commit: https://github.com/AutoMapper/AutoMapper/commit/7930dd32e61aa101ff92df5b2cde989e10ac5c0c

We have common libraries used by apps in NET6 and also large project in NET Framework.
Now we have to multitarget those libraries if we want to use latest version of AutoMapper.

**Comment by jbogard** [comment]
I don't use full .NET anymore and haven't for many years now.

**Reply by Karql** [reply]
@jbogard i do not think is the right direction.

AutoMapper is wiedly used. There are many legacy projects that still use NET Framework.
Or like in my case libraries referenced both by NET Core and NET Framework.

In linked commit there is only one thing from netstandard2.1: `GetValueOrDefault`

IMO when is no real need to update to netstandard2.1 it would be better to stay with netstandard2.0.

**Reply by jbogard** [reply]
I don't want to target `net461`, that's the main issue. I don't have it on my machine. @lbargaoanu do you see any issues just targeting `netstandard2.0`? I don't want to include tests for full .NET either. Just target latest .NET versions.

**Reply by lbargaoanu** [reply]
Targeting `netstandard2.0` is a step backward. We already use APIs that are not there. And, of course, people will use it on any desktop framework where the standard is not properly implemented and we're back to the old problems with value tuples and whatever. For us, as maintainers, it's definitely an improvment. After all, it's the same thing they did with `.Net Core`. We dropped a few things and we made some performance improvements. The reach seems wide enough to me. Other than that, 10.1.1 works just fine. If someone wants to maintain a 461 version, they're welcome to do so.

**Reply by Karql** [reply]
Support `netstandard2.0` would be great.

**Reply by jbogard** [reply]
Ah right, that's the main issue. We had to add the `net461` target in addition to the `netstandard2.0` one because of the reasons brought up [here](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting). So while we'd make some people happy with `netstandard2.0`, it would open the door to problems with anything lower than `net472`.

**Reply by Karql** [reply]
It was officially told that library in `netstandard2.0` may not works properly on `net461`.

![image](https://user-images.githubusercontent.com/173098/151381193-75dce4f0-ae43-41d7-898d-391e7347e942.png)

You can targeting just `netstandard2.0` without `net461`.

Targeting both is only suggestion:
![image](https://user-images.githubusercontent.com/173098/151388130-5789a4d0-5d5f-4bca-9b6e-72878ee638cd.png)

But by targeting `netstandard2.0` instead of `netstandard2.1` you will not cut pepole that use `NET Framework` from next cool features that you will add in the future.

> We already use APIs that are not there.

As I can see there are only:
- `GetValueOrDefault`
- `TryAdd`

I think revert deleting those extensions is not a big deal.

Best regards,
Mateusz

**Reply by Karql** [reply]
I've just tried to targeting back `netstandard2.0` only those changes are necessary:

https://github.com/Karql/AutoMapper/commit/1039f944cd5cb1fa487680d90095982b3fb01781

All tests are green:
![image](https://user-images.githubusercontent.com/173098/151386240-35319ff2-63d2-49e9-8604-089b878ff492.png)

I don't think this is step backward.

**Reply by jbogard** [reply]
The problem is if we target .NET Standard 2.0, we'll also get asked to target .NET 4.6.1 because of the problems called out in that. We already had to do that once, add the `net461` target. That's really hard to avoid if you target only `netstandard2.0`. There's no `netstandard2.0.1` that only targets `net472` and above.

**Reply by Karql** [reply]
Yes I understand it but as I said you can only targeting `netstandard2.0` and mark it in documentation or other place that `NET Framework >= 4.7.2` is required.

Exactly the same has been done in e.g. EF Core 3.

![image](https://user-images.githubusercontent.com/173098/151490549-27a11a36-3338-4ee8-9a64-08ab649c82be.png)

![image](https://user-images.githubusercontent.com/173098/151490603-ed7c942b-8430-4498-8abe-c835e3713616.png)

They changed to `netstandard2.1` in 3.0 version. There was a big feedback there is too soon etc. Then in 3.1 they reverted to `netstandard2.0`.
https://github.com/dotnet/efcore/issues/18141#issuecomment-542798580

IMO considering case with `net461` that you described it is much better to target `netstandard2.0` and let AutoMapper to work with `NET Framework >= 4.7.2` than drop support at all.

Best regards,
Mateusz

**Reply by jbogard** [reply]
Yes but EF Core 5.0 does not. I don't use .NET 4.x. I have no need or reason to support it. Neither do any of my clients that use AutoMapper. I don't want to even test on 4.7.2.

If a client or sponsor wanted to pay for supporting 4.7.2 then fine, I'd support it. Otherwise I don't have any need for it.

**Reply by martyt** [reply]
Removing .NET Standard 2.0 support from AutoMapper is going to have a cascading negative effect that you may not have considered 

For example, we have a .NET 6 application that uses `IdentityServer4.EntityFramework.Storage` (among other Identity Server 4 packages), which still targets .NET Standard 2.0, hasn't been updated in a year or so and may not be anytime soon (because the code is stable). That NuGet package also depends on AutoMapper but - as you can guess - since the package targets .NET Standard 2.0, users of that package are precluded from upgrading to AutoMapper 11 even though we don't have a .NET Framework application anywhere in sight.

In my experience, adoption of .NET Standard 2.1 isn't terribly widespread; .NET Standard 2.0, however is broadly used. 

If there's no harm (and I don't think there is) in adding a .NET Standard 2.0 target to AutoMapper, I really wish you'd consider it. Otherwise there are going to be lots of people who will be stuck at 10.1.1 forever.

**Reply by jbogard** [reply]
Identity Server 4 support ends November 2022. Its successor, Duende.IdentityServer, targets `net6.0` exclusively.

If folks want a version that targets full .NET, they can use 10.1.1.

If I targeted .NET Standard 2.0, I'd have to have a big red warning label "only .NET 4.7.2 and later". Which I'd rather not, I already went that path and it was problematic enough we added the `net461` target.
